### ISSUE TITLE
RFE: handle corrupted /etc/machine-id nicer (e.g. warn loudly and overmount temporarily with a valid one)

### ISSUE THREAD

**Issue by vedranmiletic** [systemd/systemd#4475]
### Submission type
- [X] Bug report
- [ ] Request for enhancement (RFE)
### systemd version the issue has been seen with

231
### Used distribution

Fedora 25
### In case of bug report: Expected behaviour you didn't see

journald working.
### In case of bug report: Unexpected behaviour you saw

```
# systemctl | grep failed
● systemd-journald.service               loaded failed failed    Journal Service
● systemd-tmpfiles-setup-dev.service     loaded failed failed    Create Static Device Nodes in /dev
● systemd-tmpfiles-setup.service         loaded failed failed    Create Volatile Files and Directories
● systemd-journald-audit.socket          loaded failed failed    Journal Audit Socket
● systemd-journald-dev-log.socket        loaded failed failed    Journal Socket (/dev/log)
● systemd-journald.socket                loaded failed failed    Journal Socket
```
### In case of bug report: Steps to reproduce the problem
1. Boot Fedora 25. 
### Notes
- Potentially applies to ARM machines only.
- [Downstream bug report.](https://bugzilla.redhat.com/show_bug.cgi?id=1387799)

**Comment by poettering** [comment]
Is your /etc/machine-id file valid?

**Comment by vedranmiletic** [comment]
With 230:

```
# cat /etc/machine-id
ace35669f0b24a84ba4de3caa96c0d44
e3d
```

With 231:

```
# cat /etc/machine-id
556f48e837bc4424a710fa2e2c9d3e3c
e3d
```

**Comment by poettering** [comment]
What happens if you invoke "SYSTEMD_LOG_LEVEL=debug systemd-tmpfiles --create --clean --remove" on the command line on a boot where this failed? Can you paste the output it generates?

**Comment by vedranmiletic** [comment]
[systemd-tmpfiles output](https://gist.github.com/vedranmiletic/395c0091a2e3561e2f328e79c88138ac)

**Comment by poettering** [comment]
Interesting. Can you rerun the same command in strace?

`SYSTEMD_LOG_LEVEL=debug strace -o log -s500 systemd-tmpfiles --create --clean --remove`

And provide the "log" file this creates?

**Comment by vedranmiletic** [comment]
[strace systemd-tmpfiles output](https://gist.github.com/vedranmiletic/0d2414a2800f7321539ce5c0db196d8d)

**Comment by poettering** [comment]
You machine ID file is corrupted, it contains two newlines.

**Comment by poettering** [comment]
I figure in cases like yours we should have refused booting, or overmounted /etc/machine-id with a transient but valid machine ID with a lout warning

**Comment by vedranmiletic** [comment]
I'm not sure how that happened. I went from Fedora 22 to Fedora 24 directly, which [should work](https://www.happyassassin.net/2016/01/13/upgrading-from-the-previous-stable-fedora-release/).

Still, systemd 230 works fine, only 231 breaks.

I would prefer a visible warning with a suggestion how to fix the problem.

**Comment by vedranmiletic** [comment]
It's not ARM-specific. It's not Fedora specific. Just reproduced it on Debian sid x86-64.

**Comment by poettering** [comment]
What do you mean by "reproduced" it? Did you generate a broken machine-id file by hand? Or did the breakage happen on its own?

**Comment by vedranmiletic** [comment]
I got the same error with systemd 231 on Debian x86_64. Breakage of the machine-id happened on it's own.

**Comment by vedranmiletic** [comment]
@poettering Any decision on this?

**Comment by evverx** [comment]
https://github.com/systemd/systemd/pull/3777/files#diff-9aad3279a6a5d7b21bac63a08a86e282L129
i.e. `sd_id128_get_machine (..v230)` ignored trailing bytes (34..37) 

Looks like a regression to me

I'm not sure. But seems like we don't overwrite the `/etc/machine-id`: we simply write 33 bytes (34..37 stays the same)

**Comment by evverx** [comment]
> But seems like we don't overwrite the /etc/machine-id

Yes, we don't overwrite the `/etc/machine-id`
The simplest way to reproduce:
```
# mkdir -p ROOT
# ./libtool --mode=execute ./systemd-machine-id-setup --root=$(pwd)/ROOT
Initializing machine ID from random generator.
# echo e3d >>$(pwd)/ROOT/etc/machine-id
# cat ROOT/etc/machine-id
3e08b21696c940d293d4a361f6dfb007
e3d

# ./libtool --mode=execute ./systemd-machine-id-setup --root=$(pwd)/ROOT
Initializing machine ID from random generator.
# cat ROOT/etc/machine-id
6fc425bf2c0d4ae68a37a810cc33c1ff
e3d
```

**Comment by mbiebl** [comment]
Fwiw, we just had another report on the debian-user mailing list, where a user had a broken /etc/machine-id containing
```
# cat /etc/machine-id
26da2c29c6a545fd9af95d29ca9b5a5a
6df
```

**Comment by evverx** [comment]
@mbiebl , the quick fix is:
```diff
diff --git a/src/core/machine-id-setup.c b/src/core/machine-id-setup.c
index c83bb56..df3cc74 100644
--- a/src/core/machine-id-setup.c
+++ b/src/core/machine-id-setup.c
@@ -146,14 +146,18 @@ int machine_id_setup(const char *root, sd_id128_t machine_id, sd_id128_t *ret) {
                 r = generate_machine_id(root, &machine_id);
                 if (r < 0)
                         return r;
+        }

+        if (writable) {
                 if (lseek(fd, 0, SEEK_SET) == (off_t) -1)
-                        return log_error_errno(errno, "Failed to seek: %m");
-        }
+                        return log_error_errno(errno, "Failed to seek %s: %m", etc_machine_id);
+
+                if (ftruncate(fd, 0) < 0)
+                        return log_error_errno(errno, "Failed to truncate %s: %m", etc_machine_id);

-        if (writable)
                 if (id128_write_fd(fd, ID128_PLAIN, machine_id, true) >= 0)
                         goto finish;
+        }

         fd = safe_close(fd);
```
Does it work for you?

Hm, `id128_write` doesn't overwrite file too. So, the `--commit` is broken...

**Comment by mbiebl** [comment]
@evverx not sure what this bug report is about, systemd-machine-id not producing a corrupt /etc/machine-id or systemd handling such a corrupt machine-id more gracefully.

**Comment by evverx** [comment]
@mbiebl , well, an ideal solution should fix both problems.
Suppose we have:
```
26da2c29c6a545fd9af95d29ca9b5a5a
6df
```
We should use `26da2c29c6a545fd9af95d29ca9b5a5a` (`v230`-compatible)
But we should remove the `6df`-part too.

**Comment by evverx** [comment]
And, actually, [this](https://github.com/systemd/systemd/issues/4475#issuecomment-255884454)
> or overmounted /etc/machine-id with a transient but valid machine ID with a lout warning

already implemented.

But `systemd-machine-id-commit.service` restores the "broken" id due to the `id128_write`-bug

**Comment by evverx** [comment]
Fix is waiting in #4877
